qt/aqt/reviewer.py

```diff
@@ -125,6 +125,51 @@ def rating_from_ease(ease: int) -> CardAnswer.Rating.V:
             return CardAnswer.EASY
 
 
+class State(Enum):
+    WAITING = 0
+    READY = 1
+    DELAYED = 2
+
+
+class EventualCallback:
+    """Enables suspension of execution until a callback is ready.
+
+    After a run singal, the callback is executed as soon as a ready signal is received,
+    which may be immediately. A ready signal is also emitted automatically a certain
+    time after a wait signal.
+    """
+
+    _state = State.READY
+
+    def __init__(self, parent: QObject, callback: Callable[[], None]) -> None:
+        self._timer = QTimer(parent)
+        self._timer.setSingleShot(True)
+        qconnect(self._timer.timeout, self.ready)
+        self._callback = callback
+
+    def wait(self) -> None:
+        """Suspend any runs until the next ready signal or timeout.
+        Must only be called on the main thread.
+        """
+        self._state = State.WAITING
+        self._timer.start(500)
```

Comment:
Timeouts might not be the best way to approach this - I can see it leading to difficult bug reports in the future where users report that their custom study settings are failing to be applied on very rare occasions, and our future selves may not realise it's caused by some temporary lag in DOM rendering.

Because we're passing data between the webview and Python even if we pass the context in, we do still need to wait - but a way to do that which should be more reliable is by evaluating the custom js with evalWithCallback() - provided any errors are caught and swallowed, that should guarantee we get a callback that tells us when the bottom section should be rendered, and we don't need to worry about timeouts (infinite loops in JS notwithstanding - a timeout doesn't help there anyway, as the webview will effectively be hung)